plugin/src/main/java/com/playmonumenta/scriptedquests/commands/Music.java

```diff
@@ -0,0 +1,298 @@
+		}
+	}
+
+	private static class PlayerState implements Runnable {
+		private static final ZoneId TIMEZONE = ZoneOffset.UTC;
+
+		private final UUID mPlayerId;
+		private @Nullable Song mNow = null;
+		private @Nullable Song mNext = null;
+		private LocalDateTime mNextTime = LocalDateTime.MAX;
+
+		public PlayerState(UUID playerId) {
+			mPlayerId = playerId;
+		}
+
+		public void playNow() {
+			Player player = Bukkit.getPlayer(mPlayerId);
+			if (player == null) {
+				cancelNext();
+				cancelNow();
+				return;
+			}
+
+			if (millisToRefresh() <= 0) {
+				cancelNow();
+			}
+
+			if (mNow == null) {
+				if (mNext == null) {
+					return;
+				}
+				mNow = mNext;
+				if (!mNext.mIsLoop) {
+					mNext = null;
+				}
+			}
+
+			mNextTime = LocalDateTime.now(TIMEZONE).plus(mNow.mSongDuration, ChronoUnit.MILLIS);
+			player.playSound(player.getLocation(), mNow.mSongPath, mNow.mCategory, mNow.mVolume, mNow.mPitch);
+			mRealTimePool.schedule(this, millisToRefresh(), TimeUnit.of(ChronoUnit.MILLIS));
+		}
+
+		public void playNow(Song song) {
+			cancelNext();
+			if (mNow != null && !mNow.equalsSong(song)) {
+				cancelNow();
+			}
+			playNext(song);
+		}
+
+		public void cancelNow() {
+			if (mNow == null) {
+				return;
+			}
+
+			Player player = Bukkit.getPlayer(mPlayerId);
+			if (player != null) {
+				player.stopSound(mNow.mSongPath, mNow.mCategory);
+			}
+			mNow = null;
+			mNextTime = LocalDateTime.MAX;
+		}
+
+		public void playNext(Song song) {
+			mNext = song;
+			if (mNow == null || millisToRefresh() <= 0) {
+				playNow();
+			}
+		}
+
+		public void cancelNext() {
+			mNext = null;
+		}
+
+		public long millisToRefresh() {
+			try {
+				return LocalDateTime.now(TIMEZONE).until(mNextTime, ChronoUnit.MILLIS);
+			} catch (ArithmeticException ignored) {
+				return Long.MAX_VALUE;
+			}
+		}
+
+		@Override
+		public void run() {
+			if (millisToRefresh() <= 0) {
+				new BukkitRunnable() {
+					@Override
+					public void run() {
+						playNow();
+					}
+				}.runTask(Plugin.getInstance());
+			}
+		}
+	}
+
+	private static final ScheduledThreadPoolExecutor mRealTimePool = new ScheduledThreadPoolExecutor(1);
+	private static final ConcurrentMap<UUID, PlayerState> mPlayerStates = new ConcurrentHashMap<>();
+
+	public static void register() {
+		List<Argument<?>> arguments = new ArrayList<>();
+		arguments.add(new MultiLiteralArgument("music"));
+		arguments.add(new MultiLiteralArgument("play"));
+		arguments.add(new MultiLiteralArgument("now", "next"));
+		arguments.add(new EntitySelectorArgument.ManyPlayers("players"));
+		arguments.add(new SoundArgument.NamespacedKey("music path"));
+		arguments.add(new DoubleArgument("duration in seconds", 0.001));
+
+		List<Argument<?>> optionalArguments = new ArrayList<>();
+		optionalArguments.add(new BooleanArgument("is loop"));
+		optionalArguments.add(new MultiLiteralArgument("master", "music", "record", "weather", "block", "hostile", "neutral", "player", "ambient", "voice"));
```

Comment:
could be simplified/generified to Arrays.stream(SoundCategory.values()).map(Enum::toString)

Reply:
Wouldn't those be all caps instead of all lowercase? I suppose I could use this and convert to lowercase, which would add a bit of complexity, but also mean I wouldn't need to adjust this if a new sound category is added.

Reply:
Just remembered there's a difference between `/playsound`'s sound category `block` and Bukkit's category `blocks` as well.